src/watermarks.ml

```diff
@@ -234,39 +237,27 @@ let get_name ~files ?name () =
   end;
   name
 
-let subst_git ?name () =
-  let rev = "HEAD" in
-  let git =
-    match Bin.which ~path:(Env.path Env.initial) "git" with
-    | Some x -> x
-    | None -> Utils.program_not_found "git" ~loc:None
-  in
-  let env = Env.initial in
+let subst ?name vcs =
   let+ ((version, commit), files) =
     Fiber.fork_and_join
       (fun () ->
-        Fiber.fork_and_join
-          (fun () ->
-              Process.run_capture Strict git ["describe"; "--always"; "--dirty"]
-                ~env)
-          (fun () ->
-              Process.run_capture Strict git ["rev-parse"; rev]
-                ~env))
-      (fun () ->
-        Process.run_capture_lines Strict git
-          ["ls-tree"; "-r"; "--name-only"; rev] ~env)
+         Fiber.fork_and_join
+           (fun () -> Vcs.describe vcs)
+           (fun () -> Vcs.commit_id vcs))
+      (fun () -> Vcs.files vcs)
   in
-  let version = String.trim version in
-  let commit  = String.trim commit  in
   let name = get_name ~files ?name () in
   let watermarks = make_watermark_map ~name ~version ~commit in
-  List.iter files ~f:(fun fn ->
-    let path = Path.in_source fn in
+  List.iter files ~f:(fun path ->
     if is_a_source_file path then
       subst_file path ~map:watermarks)
 
 let subst ?name () =
-  if Sys.file_exists ".git" then
-    subst_git ?name ()
-  else
-    Fiber.return ()
+  match
+    Sys.readdir "."
```

Comment:
How about using `Path.readdir_unsorted` here? It's probably overkill, but it will make the error a bit nicer if the user messed up permissions somehow.

Reply:
Hmm, I'm not sure is completely necessary here. In the case of the file_tree, it seems important as we are reporting an error for a directory somewhere far from where the user ran dune, so it's not obvious why dune is not happy. In this case, we are reporting an error for the current directory and it's going to be no surprise that dune complains if the current dir is not readable.

That said, we should probably have a special case for `Sys_error` in report_error.ml, so that we don't print `exception Sys_error(...)` which is indeed ugly.

Reply:
Okay, that would be acceptable as well.